### DISCUSSION TITLE
Angular Material CDK Table with storybook

### DISCUSSION THREAD

**Discussion by Velemir** [storybookjs/storybook#25352] [Help] [answered]
### Summary

Hello!

I'm trying to add a story to my angular material table implementation, I'm using MatTable with angular 14, cdk 14, angular material 14.

The problem is that it always gives me an error with dependencies, I created a component that is a wrapper and extended this with CdkTable, for the component I added providers such as 

providers: [
  { provide: CDK_TABLE, useExisting: MyTableComponent },
  { provide: _VIEW_REPEATER_STRATEGY, useClass: _DisposeViewRepeaterStrategy },
  { provide: _COALESCED_STYLE_SCHEDULER, useClass: _CoalescedStyleScheduler },
  { provide: STICKY_POSITIONING_LISTENER, useValue: null },
]

moduleMetadata({
  imports: [MatTableModule, MyTableModule]
})

but there is still a problem with dependencies over myTable component

Do you have any suggestions or event better, examples how to show wrapper for angular cdk table in storybook?

**Comment by Marklb** [comment]
Based on the limited information, I think this is more of a question for Angular CDK than it is Storybook, but with more information someone here may be able to help.

It would be ideal, if you could provide a reproduction. The minimal pieces of code and very brief descriptions, that you provided, don't really explain what you are trying to do. So, the best I can do is guess.

I assume you are using Storybook 6.5, so I forked one of my old Storybook 6.5 and Angular 14 Stackblitz projects. The following example is my best guess at what you are trying to do, but if not then you could fork that project as a starting point for a reproduction of what you are doing.
I have never tried extending cdk-table, but your description sounds like that is what you are doing, so I did that in my example. The way I did it may not be correct, since I just quickly threw it together, but it seems to work. You mentioned MatTable, but I don't see anything about how you would be using it, so I just left that dependency out of my example and only imported the CdkTable. If MatTable is important to your question then you would need to add it, if you fork my Stackblitz. https://stackblitz.com/edit/github-phpyee-8pkekz?file=src%2Fapp%2Fexample%2Fexample.component.ts,src%2Fapp%2Fexample%2Fexample.stories.ts

If you provide a reproduction then the following may not be necessary, but below are some questions I have before I would really be able to say that I understand your question:

1. You mention an error with dependencies. What is the error that you are referring to? Also, is it an error installing dependencies? Is it an error compiling? Is it an error thrown at runtime?
2. You mention that your component is a wrapper and you extend CDKTable. What are you wrapping? Are you wrapping CdkTable, some other component or calling your component, that extends CdkTable, a wrapper?
3. You mention adding providers for the component. Does that mean you put them in the component's `@Component({...})` decorator or did you provide them somewhere else for the components injector to inherit from, like a parent NgModule?
4. I assume you added the `moduleMetadata(...)` snippet to your story, but how does the Story look? Do you let Storybook autogenerate the template or do you define a template? I assume you define a template for the story, since a cdk-table is a little too unique for Storybook to autogenerate, unless you are using another component that uses the cdk-table/your component in it's template. There are multiple valid ways your Story could be written, so I can only guess.

**Reply by Velemir** [reply]
Thank you for your reply, your example is perfect to check. I'm sorry for giving so little details.

I'm using storybook 7.4 and the error I receive is Error: NG0202 with dependencies for Table component.
I will check what you prepared and get back to you with more examples from my side.

**Reply by Velemir** [reply]
Hello, 

I'm sorry for not replying for couple days, I was trying to figure out what's wrong with my setting and why it's working on your side, then I limited the test case to just this package.json setting when I finally reproduced my issue

  "dependencies": {
    "@angular/animations": "~14.3.0",
    "@angular/cdk": "^14.2.7",
    "@angular/common": "~14.3.0",
    "@angular/compiler": "~14.3.0",
    "@angular/core": "~14.3.0",
    "@angular/forms": "~14.3.0",
    "@angular/platform-browser": "~14.3.0",
    "@angular/platform-browser-dynamic": "~14.3.0",
    "@angular/router": "~14.3.0",
    "rxjs": "~7.5.6",
    "tslib": "^2.4.0",
    "zone.js": "~0.11.6"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~14.2.12",
    "@angular/cli": "^14.2.12",
    "@angular/compiler-cli": "~14.3.0",
    "@babel/core": "^7.18.5",
    "@compodoc/compodoc": "^1.1.19",
    "@storybook/addon-actions": "^7.5.3",
    "@storybook/addon-essentials": "^7.5.3",
    "@storybook/addon-links": "^7.5.3",
    "@storybook/angular": "^7.5.3",
    "@storybook/builder-webpack5": "^7.5.3",
    "@storybook/manager-webpack5": "^6.5.16",
    "babel-loader": "^8.2.5",
    "typescript": "~4.7.2"
  }

The problem appears only when you do build-storybook and try to run index.html. When you run local server it looks fine, but when I wanted to publish my storybook page, it happened. 

Thank you in advance for any help.

**Reply by Velemir** [reply]
So actually, only this makes me trouble 

    "@storybook/addon-actions": "^7.5.3",
    "@storybook/addon-essentials": "^7.5.3",
    "@storybook/addon-links": "^7.5.3",
    "@storybook/angular": "^7.5.3",

**Reply by Marklb** [reply]
I still can't reproduce any problem. Did you do an upgrade from Storybook 6 to 7 or an upgrade of Angular recently? In the dependencies that you posted, I see:
```json
"devDependencies": {
  ...
  "@storybook/builder-webpack5": "^7.5.3",
  "@storybook/manager-webpack5": "^6.5.16",
  ...
}
```
I don't think that would have anything to do with the error that you are getting, but Storybook versions should not be mixed and it makes me wonder if something else in your configuration could be the issue.

I still can only guess what error that you are seeing, because I can't reproduce anything and you haven't shared an error output. Could you not reproduce it by modifying my Stackblitz or creating your own project that you can share?

**Reply by Marklb** [reply]
I forked my Stackblitz and used the dependencies that you provided, to show that I can't reproduce it with them. https://stackblitz.com/edit/github-phpyee-mzsje7?file=src%2Fapp%2Fexample%2Fexample.component.ts,src%2Fapp%2Fexample%2Fexample.stories.ts,package.json

**Reply by Velemir** [reply]
I reproduced the problem on 2 laptops and a pc :D

What I did is I downloaded your project from stackblitz, 
then I did:
- npm install 
- npm run build-storybook

in folder storybook-static I launched index.html with live-server (or whatever you want to run index.html as localhost)

![image](https://github.com/storybookjs/storybook/assets/6141826/4837cf42-e59f-4be6-af23-ee7c12e60b96)

![image](https://github.com/storybookjs/storybook/assets/6141826/81c8171a-8a59-4273-9903-10eaf9e8ee2b)

The problem only occurs when you build storybook, not in live preview.
Hope it helps to reproduce the scenario, it looks like SB 7 requires some additional configuration for CDK Table since SB 6

**Reply by Marklb** [reply]
I haven't had a chance to really debug this, since it is a little more difficult to debug a static build, but  I can reproduce the error. I should have more time to debug it later this week, when I finish a project I am busy with for work.

**Reply by Marklb** [reply]
After reproducing the error, I thought it looked familiar. This seems to be a known issue that hasn't been fixed, but has a workaround. #23534

**Reply by Velemir** [reply]
You are right, it might be the same problem, I'm not sure where to put "enableProdMode": false, I would appreciate if you could show me working example on stackblitz.

Also thank you for taking time to support me on this problem.

**Reply by Marklb** [reply, marked as answer]
The documentation for those options need to be updated, but it just goes in the options for the `"@storybook/angular:build-storybook"` builder. So, in your `angular.json`, you should already have an architect object using that builder and you just add `"enableProdMode": false` to it's options.
```json
{
  ...,
  "build-storybook": {
    "builder": "@storybook/angular:build-storybook",
    "options": {
      ...,
      "enableProdMode": false
    }
  },
  ...
}
```

**Reply by Velemir** [reply]
That works great! thank you so much!
